Fixes #12718 - smart proxy log table

lzap opened the pull request:
This patch adds a table with logs fetched from Smart Proxy /logs new API. It
returns latest N log entries from a simple memory buffer. The table supports
filtering by level, custom filtering and it introduces several overview
status cards with error/warning counts, failed modules list with latest
known error messages.

Do not merge before Smart Proxy API is merged:

https://github.com/theforeman/smart-proxy/pull/358

TODO:
- [x] Implement Expire All button (only filters seen logs not delete them really)
- [x] Use dropdown for filter rather than Y icon.
- [x] Data table loading should be AJAX instead hardcoded
- [x] Move status cards to Overview or Services tab
- [x] Remove temporary JS after Puppet PR is merged
- [x] Change tab when failed module is clicked
- [x] Merge Smart Proxy API PR

lzap committed:
Fixes #12718 - smart proxy log table

This patch adds a table with logs fetched from Smart Proxy /logs new API. It
returns latest N log entries from a simple memory buffer. The table supports
filtering by level, custom filtering and it introduces several overview
status cards with error/warning counts, failed modules list with latest
known error messages.

ohadlevy commented:
@lzap can you attach some screenshots / gifs ?

lzap commented:
Failed modules is a list, clicking on it will bring you to the Log tab with word "module" filtered out from messages. Hovering on it will show you the latest initialization error. That's all the current API can do at this point.

![image](https://cloud.githubusercontent.com/assets/49752/12445512/dd16fb5a-bf64-11e5-81f0-5e966545e0d7.png)

The logs tab, filtering via text field, clicking on the Filter icon cycles through _ERROR -> WARNING -> INFO/DEBUG -> All_ (hover for a tip).

![image](https://cloud.githubusercontent.com/assets/49752/12445552/139bb2a6-bf65-11e5-8498-00c400952544.png)

Status cards are clickable:

![image](https://cloud.githubusercontent.com/assets/49752/12445564/37c87272-bf65-11e5-9e5f-21e51cb45d82.png)

And the log detail:

![image](https://cloud.githubusercontent.com/assets/49752/12449424/740ae1b2-bf7d-11e5-9861-269cac5071e8.png)

lzap commented:
Feel free to start with review, I already rebased locally but I won't push until the initial round of comments.

ohadlevy commented:
@lzap so just to be clear, there is no per plugin logs right?

lzap commented:
In general, no. But we collect last initialization error and list of plugins which failed. That's presented on the first screenshot (Disabled Features). Hover over to see the error message, click to filter out "pluginname" from the log table.

ohadlevy commented:
would that allow a plugin (e.g pulp) to show that error message instead of
the disk usage bar (for example)?

lzap commented:
Sure, let's discuss here how you prefer me to present this, this is just the beginning. But as you can see, the first two tabs are rather empty, I have nothing to hook into right now :-) I will rebase once folks merge their work.

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -88,6 +88,27 @@ def has_feature?(feature)
     self.features.any? { |proxy_feature| proxy_feature.name == feature }
   end
 
+  def logs_buffer
```
why here and not in its own status objects?

ohadlevy commented on app/models/smart_proxy.rb:
```diff
@@ -88,6 +88,27 @@ def has_feature?(feature)
+    logs_buffer.each do |log|
+      level = log['level'].try(:downcase)
+      if aggregated[level]
+        aggregated[level] += 1
```
would it make more sense for the proxy api to return those counters?

ohadlevy commented on app/services/proxy_status.rb:
```diff
@@ -16,7 +19,14 @@ def api_versions
   end
 
-  # TODO: extract to another status implementation
+  def logs
```
this needs rebase

lzap commented:
Pushed rebased version, Status classes is good point, I was thinking this is too simple that we don't need it. For the aggregation, I do not want to make the logs API complex, I think this is view's job to aggreggate things.

lzap commented:
I moved status card to the Overview page and refactored the code so it's all AJAX. Both cards, disabled features and the table are fetched via AJAX now.

@tbrisker question - when the proxy is offline, my AJAX responds with 500 and error message in plain text, but I don't see the fancy red error box. Instead, my spinner keeps rotating forever. How do I get the same behavior you have in your patch? ;-)

tbrisker commented:
@lzap It should once my patch gets in, take a look at:
https://github.com/theforeman/foreman/pull/3018/files#diff-55c5b7aecfb519d0e4880eaf2788eb6eR282
https://github.com/theforeman/foreman/pull/3018/files#diff-6d0fe7f9c2b2b2974d973cd2d242aea4R1

tbrisker commented on app/views/smart_proxies/logs/_list.html.erb:
```diff
@@ -0,0 +1,50 @@
+    'WARN' => 'W',
+    'ERROR' => 'E',
+    'FATAL' => 'E'
+  }
```
I would move these to a helper.

tbrisker commented on app/views/smart_proxies/logs/_list.html.erb:
```diff
@@ -0,0 +1,50 @@
+        </tr>
+    </thead>
+    <tbody>
+      <% @smart_proxy.statuses[:logs].log_entries.each do |log| %>
```
I would suggest passing `:locals => {:log_entries => @smart_proxy.statuses[:logs].log_entries}` from the controller.

lzap commented:
Ok now I see nicer errors there, but the formatting is a bit weird. I am not able to switch over to the Logs tab.

![image](https://cloud.githubusercontent.com/assets/49752/12550651/7aa116b8-c365-11e5-8659-f1598276cc9c.png)

lzap commented:
Scratch my comment above, got unclosed div after rebase, it looks good.

lzap commented:
Rebased, implemented all checkboxes in the summary of the PR including the "Expire" button. This could perhaps go into actions drop-down, if I only knew how can I rewrite this

```
link_to_function _('Expire logs'), "expireLogs(this);", {:"data-url" => expire_logs_smart_proxy_path(:smart_proxy_id => @smart_proxy)}
```

into `display_link_if_authorized` form. This is now feature complete, I will work on tests tomorrow.

One more thing, by default we show all logs, since logs can be unclean for end-users, we might consider increasing default level to ERROR when the page opens. I can even make this a global setting if needed.

lzap commented:
Fixed unit tests, moved Expire button to dropdown, please comment. Couple of updated screenshots:

![image](https://cloud.githubusercontent.com/assets/49752/12581759/ff729e06-c438-11e5-8d30-5316f7923837.png)

![image](https://cloud.githubusercontent.com/assets/49752/12581782/202796c4-c439-11e5-8b25-2c39b23e5a54.png)

![image](https://cloud.githubusercontent.com/assets/49752/12581789/2ab4d5c0-c439-11e5-9458-c33d0be3ede1.png)

![image](https://cloud.githubusercontent.com/assets/49752/12581795/37c1bfda-c439-11e5-83ec-c1ecba71f331.png)

After I click on Expire Logs dropdown button:

![image](https://cloud.githubusercontent.com/assets/49752/12581811/4d7ce66a-c439-11e5-876f-60b4e17d9837.png)

tbrisker commented:
Would it be possible to only display these tabs/cards for proxies that support logs? Possibly add "logging" to the proxy features? right now an old version of the proxy looks very bad:
![screenshot from 2016-01-26 17-03-29](https://cloud.githubusercontent.com/assets/1540531/12584149/c5854ab6-c44e-11e5-9414-fc4b099feba0.png)

dLobatog commented:
@lzap A few minor comments at first glance:
- Cards are missing the accented blue on top - see https://rawgit.com/patternfly/patternfly/master/tests/cards.html, cards this size should have that ![screenshot from 2016-01-26 15-59-31](https://cloud.githubusercontent.com/assets/598891/12584028/d5f34b4a-c445-11e5-9a4c-0058212cacfc.png)
- I'm not sure what the last column is supposed to mean? It doesn't have a title and I don't see it on any log line ![screenshot from 2016-01-26 15-58-51](https://cloud.githubusercontent.com/assets/598891/12584042/ee0fad36-c445-11e5-8404-ce69dd76ddb1.png)
- I don't think presenting the failed cards and disabled features like this is a good idea. Just don't show anything, the error can be seen when you hover over the status icon?
  ![screenshot from 2016-01-26 15-58-15](https://cloud.githubusercontent.com/assets/598891/12584117/4405afba-c446-11e5-8c03-8517b6f9ef5c.png)
- How does this play with old versions of the smart-proxy that do not have this feature?

Thanks! It's definitely going to be of great help for debugging issues in proxies that are not easily accessible.

ohadlevy commented:
:+1: for log feature, i want to be able to disable it in the configuration
file as well.

dLobatog commented on app/assets/javascripts/proxy_status.js:
```diff
@@ -123,8 +123,78 @@ function filterCerts(state) {
   filterCerts(__('valid')+'|'+__('pending'));
 }
+
+function filterLogsReset() {
```
nitpick, this is really proxy logs, not status, mind to move to its own file?

lzap replied:
Hmmm but Puppet and Puppet CA share the same file. Should I break them all?

dLobatog commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -85,6 +85,40 @@ def destroy
 
+  def log_pane
+    render :partial => 'smart_proxies/logs/list', :locals => {:log_entries => @smart_proxy.statuses[:logs].log_entries}
+  rescue Exception => e
```
Rescue any Exception? I'd say just rescue [StandardError instead](https://robots.thoughtbot.com/rescue-standarderror-not-exception), but if you could compile a list of possible exceptions that'd be much better. I realize you're showing the error in process_ajax_error, but that can inadvertently hide long backtraces.

dLobatog replied:
Also all these methods should have a test

lzap replied:
Changing to 

```
  rescue Foreman::Exception => exception
```

Adding tests.

dLobatog commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -98,11 +132,25 @@ def requested_data
     render :json => {:success => false, :message => exception.message} and return
   end
 
+  def logs_card_data
```
This looks like an object rather than a method with a hash. Also you can test that more easily if you want (now or later)

lzap replied:
I will refactor it into class.

dLobatog commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -49,12 +50,30 @@
+          <div class="col-md-8">
+          <% if authorized_via_my_scope(:smart_proxy, :view_smart_proxies, @smart_proxy) %>
+            <div class="tab-pane" id="failed-modules"
+              data-ajax-url="<%= failed_modules_smart_proxy_path(:smart_proxy_id => @smart_proxy) %>"><%= spinner %></div>
```
Why would I want to see the disabled features?

ohadlevy replied:
this feature was started this way - I want to know that we have plugins
that were enabled but failed to start (and see the reason why).

dLobatog commented on test/test_helper.rb:
```diff
@@ -41,6 +41,8 @@
 
   Capybara.javascript_driver = :poltergeist
 
+  Capybara.current_session.driver.header('Accept-Language', 'en')
```
I don't think this should change anything, does it?

lzap replied:
That's right, it's leftover, it did not help at all.

dLobatog commented on app/controllers/api/v2/smart_proxies_controller.rb:
```diff
@@ -59,12 +59,15 @@ def refresh
+        render_error :custom_error, :status => :unprocessable_entity, :locals => {:message => e.message}
+      end
+
+      def logs
```
Untested

lzap replied:
This has tests actually. In `test/functional/api/v2/smart_proxies_controller_test.rb`. I also fixed version test and the interface.

dLobatog commented:
Left a few comments inline about untested methods and other stuff, but overall :+1: . I still have to test with an older version of the proxy.

I don't know what's going on with proxy statuses but all PRs in this area tend to have cleaner code than in other areas, something to study...

ohadlevy commented:
@dLobatog I think its clear why... :)

lzap commented:
First of all, log buffer is not a smart proxy feature, it's part of the core as it needs to initialize early. I can take a look tomorrow if I can refactor the buffer API at least to be an ordinary feature, but given amount of time, we might consider this doing it in another patch. Success scenario does work. Thanks for other comments, will work on them tomorrow.

tbrisker commented:
@lzap I would suggest perhaps just adding the 'logging' string to the proxy's feature reply at first, just so we know that the proxy supports it and don't display all the errors for proxies that are older and don't support it yet. A second step would be to allow control in the proxy settings to enable/disable the logging feature. I agree it's not quite a feature like other features in that it doesn't provide any service to hosts and is actually part of the core of the proxy, but we are maybe just used to features being "things this proxy provides to hosts" rather that "what does this proxy know to do?" - we could in the future add more features to the proxy that are aimed at the foreman or admin rather the hosts, such as (just off the top of my head) integration with various message queues or monitoring tools.

lzap commented:
@tbrisker done - https://github.com/theforeman/smart-proxy/pull/367

I am incorporating all the comments above now.

lzap commented:
Fixed @tbrisker comment about feature, it's not proper feature and also using the "plugin" framework similarly to Puppet and CA.

@dLobatog do you know how to add the missing top line to the cards? I used instructions from PatternFly yet I don't see it... :-(

lzap commented:
Ok thanks for your initial comments, current todo list is:
- [x] - introduce Logs feature and use it properly
- [x] - break proxy_status.js into separate files
- [x] - fix generic exceptions
- [x] - add tests for AJAX calls
- [x] - refactor log buffer data into class
- [x] - fix UX line for logs cards
- [x] - get rid of the hidden column

tbrisker commented:
@lzap add the `card-pf-accented` class to get the line.

lzap commented:
Pused - All comments done, except missing tests for the three new AJAX calls, working on them. Make sure to use the Smart Proxy PR from above when testing this!

dLobatog commented:
@lzap Still see a dangling column at the end of the logs list, what is it? ![](https://cloud.githubusercontent.com/assets/598891/12584042/ee0fad36-c445-11e5-8404-ce69dd76ddb1.png)

lzap commented:
Oh I forgot this one. It's a hidden column that holds level in plain text so I can filter it. The ordinary level column contains HTML (A tag) that prevents from direct filtering. I don't see it here. I need to find out a way to workaround this. I guess I could be able to use regexp instead.

lzap commented:
Rebased, added tests, the "hidden" column is gone.

lzap commented:
Fixed permission tests, API change was merged!

tbrisker commented on app/views/smart_proxies/logs/_list.html.erb:
```diff
@@ -0,0 +1,31 @@
+            data-level="<%= log['level'] %>"
+            data-message="<%= log['message'] %>"
+            data-backtrace="<%= log['backtrace'] %>">
+            <%= log['level'] %></td>
```
Should the level be translated? If so, then the filters will also need to be changed.
Also, you swapped the order of the `</td></a>`

lzap replied:
I don't think we should translate log level, we also do not translate log messages in Foreman or Proxy. IMHO this is the threashold when we should stop translating. Someone could be for example grepping logs for additional info. Fixed tag order.

tbrisker commented on app/assets/javascripts/proxy_status/logs.js:
```diff
@@ -0,0 +1,77 @@
+        var formatted;
+        try {
+          formatted = new Date(data * 1000).toLocaleString();
+        } catch (e) {
```
Why do we need this? `toLocaleString` is currently supported on all browsers.[[1]](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString#Browser_compatibility)

lzap replied:
Removed, thanks.

tbrisker commented on app/assets/javascripts/proxy_status/logs.js:
```diff
@@ -0,0 +1,77 @@
+    var modal = $(this);
+    var datetime = new Date(link.data('time') * 1000);
+    try {
+      modal.find('#modal-bt-timegmt').text(datetime.toGMTString());
```
`toGMTString` is deprecated. Please use `toUTCString` instead. [[1]](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toGMTString)

lzap replied:
If you say so... ;-)

tbrisker commented on app/assets/javascripts/proxy_status/logs.js:
```diff
@@ -0,0 +1,77 @@
+    try {
+      modal.find('#modal-bt-timegmt').text(datetime.toGMTString());
+      modal.find('#modal-bt-time').text(datetime.toLocaleString());
+    } catch (e) { /* ignore for legacy browsers */ }
```
Again, not sure we need to support ancient browsers.

tbrisker commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -28,6 +28,10 @@ def feature_actions(proxy, authorizer)
       actions << display_link_if_authorized(_("Import subnets"), hash_for_import_subnets_path(:smart_proxy_id => proxy))
     end
 
+    actions << link_to_function_if_authorized(_('Expire logs'), "expireLogs(this);",
```
This should also only show if the proxy supports this feature.

lzap replied:
Good catch.

tbrisker commented on db/seeds.d/11-smart_proxy_features.rb:
```diff
@@ -1,5 +1,5 @@
 # Proxy features
-[ "TFTP", "DNS", "DHCP", "Puppet", "Puppet CA", "BMC", "Realm", "Facts" ].each do |input|
+[ "TFTP", "DNS", "DHCP", "Puppet", "Puppet CA", "BMC", "Realm", "Facts", "Logs" ].each do |input|
```
Does it make sense to also add this via migration?

lzap replied:
Nope, we usually add just seed, as in for example a2523ccf030169eaf170c329e2f68193777ab2cb

tbrisker commented on app/views/smart_proxies/logs/_modal.html.erb:
```diff
@@ -0,0 +1,40 @@
+<div class="modal fade" id="backtrace-modal" tabindex="-1" role="dialog" aria-labelledby="bt-modal-label">
```
you can use the modal layout here instead of repeating all the modal boilerplate, take a look at https://github.com/theforeman/foreman/blob/develop/app/views/common/_modal.html.erb and https://github.com/theforeman/foreman/blob/develop/app/views/autosign/_form.html.erb#L2 for an example

lzap replied:
Changed. How about this formatting?

![image](https://cloud.githubusercontent.com/assets/49752/12639319/fc23bd04-c5a3-11e5-9402-bdb784e7838b.png)

I don't want to make the dialog wider, it is set to this size for a reason. And log lines are all long. But in the new layout, there is much more room for those.

tbrisker replied:
We do have wider modals in some places, host interfaces is the first that comes to mind.
The new layout does look better though. I think you could fit the first three lines (time and level) in one row possibly, leaving more space for the details. Also, possibly add a slight border to the overflowing areas so it would be clearer what the scroll bars are related to.

tbrisker commented on app/views/smart_proxies/logs/_modal.html.erb:
```diff
@@ -0,0 +1,40 @@
+          </tr>
+          <tr>
+              <td class="nbsp"><%= _('Message') %></td>
+              <td><div class="pre-scollable-xy-modal" id="modal-bt-message"/></td>
```
This is too small for any substential message, consider increasing the modal size:
![screenshot from 2016-01-27 23-58-21](https://cloud.githubusercontent.com/assets/1540531/12629919/a24b7976-c552-11e5-9ee4-b66087fe1bd8.png)

lzap replied:
See above comment, let me push the new layout and disscuss it.

tbrisker commented on app/views/smart_proxies/logs/_errors_card.html.erb:
```diff
@@ -0,0 +1,18 @@
+<div class="card-pf card-pf-accented card-pf-aggregate-status card-pf-with-action">
+  <h2 class="card-pf-title">
+    <%= n_('%s info message', '%s info messages', logs.info_messages) % logs.info_messages %>
```
This should link to the unfiltered logs

lzap replied:
Yeah adding.

tbrisker commented:
@lzap Looks good for the most part, just a few minor comments inline and below.
When clicking on 'expire logs' when in the logs page, the table columns seem to get resized, possibly some datatables weirdness? 
The disabled features list doesn't seem to display core modules that failed to load, but I think that is a problem on the proxy side.
Maybe add a refresh logs button on the logs page that only reloads the data - which could be useful when trying to debug without refreshing the entire page.
Also take a look at the failing tests.
Other then that I think we are good to go.

lzap commented:
Ok thanks, will work on the comments above today.

I noticed resizing, to be honest I don't know what is going on. My table should be grid 1+1+10 but apparently the top-left column sizes are larger - this does not look like a one grid "point" at all. I see you have fixed sizes, how did you chieve that with datatable?

Will take a look on failed proxy core module. Refresh button is a good idea TODO items:
- [x] incorporate comments above
- [ ] table is resized after Expire action
- [x] failed core modules not reported
- [x] add Refresh button into the Logs table
- [x] fix failing tests

lzap commented:
I just tested with TFTP module and it reports correctly, what is your use case @tbrisker ? Show me the reproducer.

tbrisker commented:
@lzap sorry, looks like it was a mistake on my side. I was getting an error:
`Couldn't enable plugin dns_nsupdate: File at '/etc/rndc.key' defined in 'dns_key' parameter doesn't exist or is unreadable`
but in fact dns was not enabled; when I enabled dns it was listed in the failed modules.

tbrisker commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -49,12 +49,31 @@
+              <%= _('Disabled Features') %>
+            </div>
+            <div class="col-md-8">
+            <% if authorized_via_my_scope(:smart_proxy, :view_smart_proxies, @smart_proxy) %>
```
I think without this permission the whole page won't show.

lzap replied:
Removed.

tbrisker commented on app/views/smart_proxies/logs/_failed_modules.html.erb:
```diff
@@ -0,0 +1,3 @@
+<% modules.each do |module_name, error_message| %>
+  <a href="#logs" onclick="filterLogsByMessage('<%= module_name %>'); setTab();" data-toggle="tooltip" data-placement="top" title="<%= error_message %>"><%= module_name.capitalize %></a>
```
You can use `Feature.name_map[module_name]` instead of `module_name.capitalize` to get the correct feature name capitalization.

lzap replied:
Done.

tbrisker commented:
Not sure what causes the table columns to resize, looks like something inside datatables. I wouldn't bother with it right now and possibly fix it in the future (or hope it goes away when migrating to datatables 1.10).

lzap commented:
Pushed all changes, except the column sizing. I am unable to reproduce, it always loads fine.

lzap commented:
I fixed the column resizing problem by adding "sWidth" into table initialization in percent size. Now both bootstrap hiding and proper size look good. Will update after your next bunch of comments :)

ohadlevy commented:
do we still need the list of features and the list of modules? IMHO we
should converge the two?

ohadlevy commented:
@lzap @tbrisker one comment about the modules in operation card:
![card](https://cloud.githubusercontent.com/assets/20167/12643105/c7081a7e-c5c2-11e5-80a6-a557256b4e63.png)

maybe swap it to something like:
![service](https://cloud.githubusercontent.com/assets/20167/12643398/9061a5c4-c5c4-11e5-8bfb-bd33008a9760.png)

lzap commented:
Sure, show me how you would like to see the list. Maybe another card? Or just remove list of features completely and show it as a hover over "4/4 features in operation"?

To your second point @ohadlevy I agree that "4/4 modules" is not ideal, I could simply show "4 features active" or just "4 Features"? I am not an UX expert, but showing an error icon with zero when everything is operating smoothly is not the best idea :-) I'd go with the success green tick icon instead.

ohadlevy commented:
right, you can easily change it, this is what i used:

``` erb
<div class="card-pf card-pf-accented card-pf-aggregate-status">
  <h2 class="card-pf-title">
    <span class="pficon pficon-service"></span>
    <a href="#">
      <span class="card-pf-aggregate-status-count">
        <%= features_total %>
      </span>
      <%= _('Features') %>
    </a>
  </h2>
  <div class="card-pf-body">
    <p class="card-pf-aggregate-status-notifications">
      <a href="#">
        <span class="card-pf-aggregate-status-notification">
          <span class="pficon pficon-error-circle-o"></span>
          <%= features_total - features_started %>
        </span>
      </a>
    </p>
  </div>
</div>
```

ohadlevy commented:
probably not possible to format via email, here is another try:

``` erb
<div class="card-pf card-pf-accented card-pf-aggregate-status">
  <h2 class="card-pf-title">
    <span class="pficon pficon-service"></span>
    <a href="#">
      <span class="card-pf-aggregate-status-count">
        <%= features_total %>
      </span>
      <%= _('Features') %>
    </a>
  </h2>
  <div class="card-pf-body">
    <p class="card-pf-aggregate-status-notifications">
      <a href="#">
        <span class="card-pf-aggregate-status-notification">
          <span class="pficon pficon-error-circle-o"></span>
          <%= features_total - features_started %>
        </span>
      </a>
    </p>
  </div>
</div>
```

lzap commented:
How about this:

![image](https://cloud.githubusercontent.com/assets/49752/12643992/095731ce-c5c0-11e5-9e0d-6979d13f64fb.png)

When there is no failed feature, there is the green icon.

lzap commented:
I also found that after Expire logs the cards are not reloaded - fixed locally.

lzap commented:
@ohadlevy @tbrisker since we are short of time, I'd rather merge the Overview page as is, once we have all the features, we can do a separate UX PR to improve overall look. We don't have all the content there yet I guess.

lzap commented:
What do you say about the new modal dialog? It's a bit simple, not sure how to position things from top to bottom in bootstrap grid, so I used `<br/>` :-)

tbrisker commented:
I still think the modal should be a bit wider, but I won't force it. You can use class `row` in bootstrap to get a new row. I still find the scrollbars with no box in the middle of the dialog a bit confusing, see for example:
![screenshot from 2016-01-28 15-41-12](https://cloud.githubusercontent.com/assets/1540531/12645893/ae23007e-c5d5-11e5-9fcb-fd4d4d707dda.png)
I'm ok with the way the cards are at the moment; if needed we can reiterate on it in the future. Seems like expiry stopped working, not sure why.

lzap commented:
Let me improve it a bit. But now do I make it wider in portable way? I don't want to specify 800px for example, that won't work on phones!

tbrisker commented:
@lzap you can set `width:1000px; max-width:80%;`  for example to make sure that it doesn't overflow on smaller screens.

ohadlevy commented:
the message can be in a `<div class=code></div>` which will be multiline and with background?

tbrisker commented:
@ohadlevy don't forget to use `` to escape html code so we can also see it :)

ohadlevy commented:
@tbrisker updated.

lzap commented:
Yeah code looks fine, basically I want to avoid textarea because I'd need to pre-allocate number of rows or do some magic around it.

Width is an issue, when I attempt to set it via `style="width:1000px; max-width:80%"` it shifts to the left:

![image](https://cloud.githubusercontent.com/assets/49752/12646369/d0b6f38a-c5cf-11e5-9d12-36258c563bbc.png)

lzap commented:
Found it!

![image](https://cloud.githubusercontent.com/assets/49752/12646650/a0ebee88-c5d1-11e5-9f23-87ad160ec5c4.png)

Not sure about the background color, I was picking up from http://getbootstrap.com/css/#helper-classes-backgrounds

lzap commented:
Pushed:
- modal is now 80 % wide (introduced `big` local to the helper)
- cards are updated when Expire logs or Refresh is called

I think we are good to go.

tbrisker commented:
@lzap Looks good! just two last comments:
1. In the modal, make the top three labels col-md-2 instead of col-md-6 so the values won't be so far off.
2. Is there a reason we don't allow the table to sort, atleast for time and level? Also, I would make the date sorting default to reverse, so newest messages show up on the first page.
Other then that, unless @ohadlevy or @dLobatog have any further comments I think we are good to go.

lzap commented:
Fixed, sorting enabled and working fine.

tbrisker commented on app/views/smart_proxies/logs/_modules_card.html.erb:
```diff
@@ -0,0 +1,14 @@
+<div class="card-pf card-pf-accented card-pf-aggregate-status card-pf-with-action">
+  <h2 class="card-pf-title">
+    <a href="#" data-toggle="tooltip" data-placement="top" title="<%= features.join(', ') %>"><%= n_('%s Features active', '%s Features active', features_started) % features_started %></a>
```
No need for this to be an `<a>` as it links nowhere which is confusing. better just a span for the tooltip or add it directly to the `<h2>`.

domcleal replied:
The singular string passed to `n_()` is plural here, shouldn't it be "Feature active" (and probably lower case, since it's the second word)?

dLobatog replied:
'Active features' even

lzap replied:
Ok changing to:

``` erb
  <h2 class="card-pf-title" data-toggle="tooltip" data-placement="top" title="<%= features.join(', ') %>">
    <%= n_('%s active feature', '%s active features', features_started) % features_started %>
  </h2>
```

lzap commented:
I hereby declare I am going to convert at least one fixtures to Factory Girl in a separate PR (as I added one example in this patch).

lzap commented:
So now it looks like:

![image](https://cloud.githubusercontent.com/assets/49752/12649554/5a936174-c5df-11e5-9c4f-39b255fef6b2.png)

tbrisker commented on app/assets/stylesheets/application.scss:
```diff
@@ -551,3 +551,23 @@ span.glyphicon.host-status {
   overflow-y: auto;
 }
+
+modal.modal-big modal-dialog.modal-big {
```
missing `.`, there is no element type `modal`

tbrisker replied:
actually just leave the .modal-big I think.

lzap replied:
Fixed.

tbrisker commented on app/views/smart_proxies/logs/_modal.html.erb:
```diff
@@ -0,0 +1,41 @@
+   :big => true,
+   :buttons => [ modal_close(_('Ok')) ]} do %>
+<div class="container">
+    <div class="row">
```
nitpick: indentation is off here (4 spaces instead of 2)

lzap replied:
Yeah.

tbrisker commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -85,6 +85,51 @@ def destroy
+    from = (params[:from].to_i rescue 0) || 0
+    if from >= 0
+      logger.debug "Expired smart-proxy logs, new timestamp is #{from}"
+      Setting[:proxy_expired_logs] = from.to_s rescue '0'
```
This means that if I expire logs on one proxy it expires them for all proxies?

lzap replied:
Bad design for sure, I will add a new column to proxy for this purpose.

tbrisker commented on app/models/smart_proxy.rb:
```diff
@@ -97,6 +97,7 @@ def statuses
       end
     end
     @statuses[:version] = ProxyStatus::Version.new(self)
+    @statuses[:logs] = ProxyStatus::Logs.new(self)
```
wouldn't this already be added by the features loop above?

lzap replied:
Leftover.

tbrisker replied:
still leftover.

lzap replied:
Oh sorry, no. This needs to be here because the generic code use `humanize` which does not work in this case (it's `Log` vs `Logs`).

lzap replied:
Wait I need to take a look.

tbrisker commented on app/services/proxy_status/logs.rb:
```diff
@@ -0,0 +1,18 @@
+
+    def logs
+      fetch_proxy_data do
+        ::SmartProxies::LogBuffer.new(api.all((Setting[:proxy_expired_logs].to_i rescue 0) || 0))
```
Again, this shouldn't be a Setting, I think it would be best to have this as an instance variable for the logs status object

lzap replied:
I am adding this to the model as status objects disappear once cache is invalidated.

tbrisker commented on app/services/proxy_status/logs.rb:
```diff
@@ -0,0 +1,18 @@
+module ProxyStatus
+  class Logs < Base
+    def api
+      ProxyAPI::Logs.new(:url => proxy.url)
```
This is already handled in the Base class

lzap replied:
Leftover, t hanks.

lzap replied:
Oh actually this and above is necessary because the class is named `Logs` while `humanize` makes it `Log` causing a name error. Since the REST API is also named `/logs`, I'd rather keep it as is.

tbrisker replied:
That's what `api_class` is for in base - I had the same problem with puppet ca.
If you take a look at the api function in base, you will notice that it actually does two things except the api call - one is reusing the api instance object if it already exists instead of creating a new one for every api call, and the second is catching any NameErrors raised by the initialization of the api object.

tbrisker commented on app/services/smart_proxies/log_buffer.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  def aggregated_logs
+    aggregated = { 'debug' => 0, 'info' => 0, 'warn' => 0, 'error' => 0, 'fatal' => 0 }
```
Why not memoize this instead of counting every time from the start?

lzap replied:
Sure!

lzap commented:
I will work on this on Monday. I think my initial idea to save it in settings was bad from day one. I need to store this in the Smart Proxy itself (in a new column). This global approach wont work across different smart proxies.

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -13,6 +13,8 @@
 //= require about
 //= require proxy_status
+//= require proxy_status/puppet
+//= require proxy_status/logs
```
why load it for all pages?

lzap replied:
How do I do this properly? We already load bunch of stuff here and I believe not loading it breaks the TurboLinks thing... Teach me please.

ohadlevy commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -1,5 +1,5 @@
 module SmartProxiesHelper
-  TABBED_FEATURES = ["Puppet","Puppet CA"]
+  TABBED_FEATURES = ["Puppet", "Puppet CA", "Logs"]
```
we should ensure they are sorted when presented (or keep the current order) as i already saw in one of the screenshots the following tabs:
Puppet, Logs, Puppet CA which makes little sense

lzap replied:
Yeah, that's caused by SQL server because this is what we do:

```
proxy.features.where('features.name IN (?)', TABBED_FEATURES).uniq.pluck("name")
```

I can add a sort (by alphanum) and we can perhaps add a priority column to features that defines the ordering later?

lzap replied:
Alphanum sorting unfortunately does not look good:

![image](https://cloud.githubusercontent.com/assets/49752/12718703/909ad5f4-c8f0-11e5-9044-af965706b240.png)

Do you want me to add a priority number to Features?

lzap replied:
I can perhaps pull this work into separate PR, let's get this one merged.

tbrisker replied:
+1 for working on this separately. The tabbed features list has room for improvements.

ohadlevy commented on app/helpers/smart_proxies_helper.rb:
```diff
@@ -28,6 +28,15 @@ def feature_actions(proxy, authorizer)
+    if proxy.has_feature?('Logs')
+      actions << link_to_function_if_authorized(_('Expire logs'), "expireLogs(this, (new Date).getTime() / 1000);",
+        hash_for_expire_logs_smart_proxy_path(:id => proxy), {
+          :"data-url" => expire_logs_smart_proxy_path(:id => proxy),
```
nitpick, you can do:

``` ruby
:data => {:url => ..., :url-errors => ...}
```

lzap replied:
Done.

lzap commented:
Rebased, any chance you take a look @tbrisker ?

tbrisker commented on app/assets/javascripts/proxy_status/logs.js:
```diff
@@ -0,0 +1,97 @@
+function filterLogsReset() {
+  var table = $('#table-proxy-status-logs').dataTable();
+  for (var i = 0; i <= 2; i++) {
+    table.fnFilter('', i);
```
You can just use `$('#table-proxy-status-logs').dataTable().fnFilter('');` instead of looping over the columns to clear all the filters.

tbrisker replied:
oh, no, never mind. seems like databales is weird.

tbrisker commented on app/assets/javascripts/proxy_status/logs.js:
```diff
@@ -0,0 +1,97 @@
+function changeFilterSelection(index) {
+  var filter = $('#logs-filter');
+  filter[0].options[index].selected = true
+  filter.select2();
```
you can just `filter.trigger('change')` instead of reinitializing selec2

tbrisker commented on app/models/setting/general.rb:
```diff
@@ -22,7 +22,8 @@ def self.load_defaults
         self.set('db_pending_seed', N_("Should the `foreman-rake db:seed` be executed on the next run of the installer modules?"), true, N_('DB pending seed')),
-        self.set('proxy_request_timeout', N_("Max timeout for REST client requests to smart-proxy"), 60, N_('Proxy request timeout'))
+        self.set('proxy_request_timeout', N_("Max timeout for REST client requests to smart-proxy"), 60, N_('Proxy request timeout')),
+        self.set('proxy_expired_logs', N_("Do not fetch proxy logs older than (UNIX epoch miliseconds)"), '0', N_('Proxy expired logs'))
```
Left over from previous version

tbrisker commented on app/views/smart_proxies/logs/_modal.html.erb:
```diff
@@ -0,0 +1,41 @@
+ :title => _("Log entry details"),
+ :big => true,
+ :buttons => [ modal_close(_('Ok')) ]} do %>
+ <div class="container">
```
nitpick: indentation

tbrisker commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -49,12 +49,29 @@
+        <% if @smart_proxy.has_feature?('Logs') %>
+          <div class="row">
+            <div class="col-md-4">
+              <%= _('Disabled Features') %>
```
I think it would be better to call this line "Failed Features", since "Disabled" might imply that they are simply set to `enable: false` in the configuration rather then failed to load.

tbrisker commented on config/initializers/1_fast_gettext.rb:
```diff
@@ -4,6 +4,9 @@
 locale_domain = 'foreman'
 
+# this prevents from reading 'en' translations in test environment
+locale_domain = 'foreman_test' if Rails.env.test?
```
Is this related somehow?

lzap replied:
Removing.

tbrisker commented on test/fixtures/settings.yml:
```diff
@@ -286,3 +286,9 @@ attributes59:
   category: Setting::Puppet
   default: "false"
   description: 'All hosts will show a configuration status even when a Puppet smart proxy is not assigned'
+attributes60:
```
Leftover

tbrisker commented:
@lzap Looks very good. There are just a few minor nitpicks and small leftovers to clean up, once they are done I'll merge this.

lzap commented:
Amended all your comments, thanks. Took the i18n hunk into separate pr.

lzap commented:
Ok pushed.

tbrisker commented:
@lzap looks like there's some weird off-by-one error in the active feature count:
![screenshot from 2016-02-03 10-37-37](https://cloud.githubusercontent.com/assets/1540531/12778313/69b14734-ca62-11e5-804c-69c15a03d6fe.png)

Other then that, looks ready to go once the tests are green.

lzap commented:
Pushed, that was miscalculation.

tbrisker commented:
Merged as 33d9f9eeec63da0322b7d318077c2a7e58ffff11, thanks @lzap!

lzap commented:
Thanks! TODO:
- some FactoryGirl coding
- introduce feature priority
